Optimize distributed.Cache.Get when lookaside cache is enabled

vanja-p opened the pull request:
There's 2 optimizations here:
1. When the entry doesn't belong in the lookaside cache (like AC responses), don't create a `teeReadCloser` in `lookasideWriter` since it's unnecessary overhead.
2. When we do write to the lookaside cache, presize the buffer

Benchmark results:
```goos: linux
goarch: amd64
cpu: AMD Ryzen 9 7900X 12-Core Processor            
                                          │    base     │               after                │
                                          │   sec/op    │   sec/op     vs base               │
GetSingle/LookasideDistPebble10/AC-24       102.2µ ± 1%   100.7µ ± 1%  -1.54% (p=0.003 n=15)
GetSingle/LookasideDistPebble10/CAS-24      1.315µ ± 1%   1.321µ ± 1%       ~ (p=0.467 n=15)
GetSingle/LookasideDistPebble100/AC-24      104.8µ ± 2%   102.3µ ± 1%  -2.44% (p=0.000 n=15)
GetSingle/LookasideDistPebble100/CAS-24     1.341µ ± 1%   1.342µ ± 3%       ~ (p=0.340 n=15)
GetSingle/LookasideDistPebble1000/AC-24     108.2µ ± 0%   105.7µ ± 0%  -2.32% (p=0.000 n=15)
GetSingle/LookasideDistPebble1000/CAS-24    107.9µ ± 1%   107.8µ ± 1%       ~ (p=0.806 n=15)
GetSingle/LookasideDistPebble10000/AC-24    144.2µ ± 1%   135.4µ ± 1%  -6.08% (p=0.000 n=15)
GetSingle/LookasideDistPebble10000/CAS-24   139.9µ ± 1%   138.0µ ± 0%  -1.37% (p=0.001 n=15)
geomean                                     38.11µ        37.47µ       -1.69%

                                          │     base     │                after                │
                                          │     B/s      │     B/s       vs base               │
GetSingle/LookasideDistPebble10/AC-24       97.66Ki ± 0%   97.66Ki ± 0%       ~ (p=0.483 n=15)
GetSingle/LookasideDistPebble10/CAS-24      7.248Mi ± 1%   7.219Mi ± 1%       ~ (p=0.479 n=15)
GetSingle/LookasideDistPebble100/AC-24      927.7Ki ± 1%   957.0Ki ± 1%  +3.16% (p=0.000 n=15)
GetSingle/LookasideDistPebble100/CAS-24     71.11Mi ± 1%   71.06Mi ± 3%       ~ (p=0.340 n=15)
GetSingle/LookasideDistPebble1000/AC-24     8.812Mi ± 0%   9.022Mi ± 0%  +2.38% (p=0.000 n=15)
GetSingle/LookasideDistPebble1000/CAS-24    8.831Mi ± 1%   8.850Mi ± 1%       ~ (p=0.783 n=15)
GetSingle/LookasideDistPebble10000/AC-24    66.16Mi ± 1%   70.44Mi ± 1%  +6.47% (p=0.000 n=15)
GetSingle/LookasideDistPebble10000/CAS-24   68.17Mi ± 1%   69.11Mi ± 0%  +1.39% (p=0.000 n=15)
geomean                                     7.929Mi        8.058Mi       +1.62%

                                          │     base      │                after                 │
                                          │     B/op      │     B/op      vs base                │
GetSingle/LookasideDistPebble10/AC-24        30.58Ki ± 0%   29.87Ki ± 0%   -2.32% (p=0.000 n=15)
GetSingle/LookasideDistPebble10/CAS-24       1.385Ki ± 0%   1.385Ki ± 0%        ~ (p=0.710 n=15)
GetSingle/LookasideDistPebble100/AC-24       30.96Ki ± 0%   30.26Ki ± 0%   -2.25% (p=0.000 n=15)
GetSingle/LookasideDistPebble100/CAS-24      1.480Ki ± 0%   1.480Ki ± 0%        ~ (p=1.000 n=15)
GetSingle/LookasideDistPebble1000/AC-24      36.46Ki ± 0%   34.21Ki ± 0%   -6.15% (p=0.000 n=15)
GetSingle/LookasideDistPebble1000/CAS-24     35.43Ki ± 0%   35.43Ki ± 0%        ~ (p=0.493 n=15)
GetSingle/LookasideDistPebble10000/AC-24    104.91Ki ± 0%   50.74Ki ± 0%  -51.63% (p=0.000 n=15)
GetSingle/LookasideDistPebble10000/CAS-24    61.12Ki ± 0%   61.09Ki ± 0%        ~ (p=0.533 n=15)
geomean                                      18.87Ki        17.00Ki        -9.93%

                                          │    base    │                after                │
                                          │ allocs/op  │ allocs/op   vs base                 │
GetSingle/LookasideDistPebble10/AC-24       482.0 ± 0%   477.0 ± 0%  -1.04% (p=0.000 n=15)
GetSingle/LookasideDistPebble10/CAS-24      23.00 ± 0%   23.00 ± 0%       ~ (p=1.000 n=15) ¹
GetSingle/LookasideDistPebble100/AC-24      482.0 ± 0%   477.0 ± 0%  -1.04% (p=0.000 n=15)
GetSingle/LookasideDistPebble100/CAS-24     23.00 ± 0%   23.00 ± 0%       ~ (p=1.000 n=15) ¹
GetSingle/LookasideDistPebble1000/AC-24     484.0 ± 0%   477.0 ± 0%  -1.45% (p=0.000 n=15)
GetSingle/LookasideDistPebble1000/CAS-24    488.0 ± 0%   488.0 ± 0%       ~ (p=1.000 n=15) ¹
GetSingle/LookasideDistPebble10000/AC-24    489.0 ± 0%   473.0 ± 0%  -3.27% (p=0.000 n=15)
GetSingle/LookasideDistPebble10000/CAS-24   484.0 ± 0%   484.0 ± 0%       ~ (p=1.000 n=15) ¹
geomean                                     226.3        224.3       -0.85%
```

vanja-p commented:
ping

vanja-p commented on enterprise/server/backends/distributed/distributed.go:
```diff
@@ -1093,34 +1127,35 @@ func (c *Cache) distributedReader(ctx context.Context, rn *rspb.ResourceName, of
+	// The median and average AC results are less than 4KiB: go/action-result-size
+	return 4 * 1024
+
+}
```
Please do. Why is gofmt not catching this?

vanja-p commented on enterprise/server/backends/distributed/distributed.go:
```diff
@@ -1093,34 +1127,35 @@ func (c *Cache) distributedReader(ctx context.Context, rn *rspb.ResourceName, of
-		buf = bytes.NewBuffer(make([]byte, 0, int(rn.GetDigest().GetSizeBytes())))
+		// Clamp the size between 0 and 10MB, to protect from invalid and
+		// malicious requests.
+		return min(10_000_000, max(0, int(rn.GetDigest().GetSizeBytes())))
```
Yep, the benchmarks are smaller than this cap. It will definitely make things slower when reads are bigger than the cap. I don't mind removing it, since we didn't have it before, but in https://github.com/buildbuddy-io/buildbuddy/pull/8887, we decided to add a cap to the pebble cache.

jdhollen commented:
Sorry for being slow here--i'm theoretically fine with this change, but there are some conflicts that you'll need to sync past now.

vanja-p commented:
> Sorry for being slow here--i'm theoretically fine with this change, but there are some conflicts that you'll need to sync past now.

I handled the conflicts and simplified this change. Turns out that implementing WriterTo here doesn't help. Also, I reverted most of the buffer size changes, but I did presize the `lookasideWriter`, since it's a trivial change.
